DOCS: added UCC User User Guide

jirikraus opened the pull request:
## What
Proposal for a UCC End User, e.g. developers of MPI applications, Guide.

## Why ?
Currently information about basic concepts like Team Layer, Collective Layer is scattered (FAQ, API Docs) and target for developers contributing to UCC or using UCC directly. Documentation for developers using UCC via e.g. an UCC enabled MPI stack is missing.

swx-jenkins3 commented:
Can one of the admins verify this patch?

Sergei-Lebedev commented:
ok to test

Sergei-Lebedev commented:
@jirikraus thanks for this PR! please change commit title to "DOCS: added UCC User User Guide" to make codestyle checker happy.

jirikraus commented:
> @jirikraus thanks for this PR! please change commit title to "DOCS: added UCC User User Guide" to make codestyle checker happy.

Thanks @Sergei-Lebedev for taking a look at this so quickly. Just updated the commit title.

manjugv commented:
@edgargabriel Can you please review?

edgargabriel commented on docs/user_guide.md:
I would suggest to remove rocm_managed from this list, we have realistically only tested with rocm device memory.

jirikraus replied:
Agree if `rocm_managed` is effectively not supported (as not tested) I agree it should be dropped. Keeping this open to address it with the other points for which I need clarification.

edgargabriel commented on docs/user_guide.md:
I am wondering whether its worthwhile adding a sentence here (or another location in the document) about the fallback path in Open MPI ucc component in case e.g. a dataype is not support by ucc. My line of thinking is to avoid a user being confused that setting the ucc variables does not show any changes for his application

jirikraus replied:
Yes agree that should be mentioned. Would this be accurate:

```suggestion
A MPI or other programming model implementation might need to execute a collective not supported by UCC,
e.g. because the datatype or reduction operator support is lacking. in these cases Open MPI does not
call into UCC but leverages its own internal collective implementation. Unless the used MPI
implementation provides facilities for this, these situations can be only detected indirectly from the
lack of UCC activity. 
```

edgargabriel replied:
I think generally its good. Two minor comments/suggestions: The second sentence, please capitalize the "in these cases" (i.e. "In these cases"). Add maybe we can add a sentence specifically how to turn on verbosity of the ucc component which will report that the fallback path is being used. E.g.

"For Open MPI, the user could set the `coll_ucc_verbose` mca parameter to increase the verbosity of the ucc collective component, and hence identify scenarios in which the ucc library is not being used."

edgargabriel replied:
> > One other item, do I recall correctly that there was also the possibility to use a configuration file instead of the environment variables? If yes, would it make sense to mention it here?
> 
> Yes I would like to add that. Where can I read up on how the file location is determined and what the format is? @edgargabriel can you comment on this and my other replies?

I think there are other people that are more knowledgeable about this topic on this list, I just recall some discussion in the ucc meeting, but I personally have not used the configuration file yet (which is why I am interested in getting a description on how precisely to create a configuration file, where it needs to be located etc.)

jirikraus replied:
@manjugv can you help here? Either directly or by making a connection?

edgargabriel reviewed (approved):
I think the document looks very good and helpful. Just a few questions, please feel free to ignore.

One other item, do I recall correctly that there was also the possibility to use a configuration file instead of the environment variables? If yes, would it make sense to mention it here?

jirikraus commented:
> One other item, do I recall correctly that there was also the possibility to use a configuration file instead of the environment variables? If yes, would it make sense to mention it here?

Yes I would like to add that. Where can I read up on how the file location is determined and what the format is?

jirikraus commented:
Thanks @Sergei-Lebedev and @edgargabriel for your reviews. Except the open item on the configuration file for which I need input (@manjugv  can you help?) I think I addressed all your comments. The specific open questions on that are:

- How is the file configuration file location is determined?
- What is the configuration file format?

manjugv commented:
> Thanks @Sergei-Lebedev and @edgargabriel for your reviews. Except the open item on the configuration file for which I need input (@manjugv can you help?) I think I addressed all your comments. The specific open questions on that are:

@shimmybalsam is working updating the FAQ and wiki for this. Do you want to cross link later or do you want to wait for it to be added?

jirikraus commented:
> > Thanks @Sergei-Lebedev and @edgargabriel for your reviews. Except the open item on the configuration file for which I need input (@manjugv can you help?) I think I addressed all your comments. The specific open questions on that are:
> 
> @shimmybalsam is working updating the FAQ and wiki for this. Do you want to cross link later or do you want to wait for it to be added?

Either way works for me. What is the ETA for the FAQ und wiki updates?

shimmybalsam commented:
> > > Thanks @Sergei-Lebedev and @edgargabriel for your reviews. Except the open item on the configuration file for which I need input (@manjugv can you help?) I think I addressed all your comments. The specific open questions on that are:
> > 
> > 
> > @shimmybalsam is working updating the FAQ and wiki for this. Do you want to cross link later or do you want to wait for it to be added?
> 
> Either way works for me. What is the ETA for the FAQ und wiki updates?

FAQ and wiki update should be done during next week

jirikraus commented:
> FAQ and wiki update should be done during next week

Then let's wait for it. @shimmybalsam can you ping me here when that is done?

shimmybalsam commented:
> > FAQ and wiki update should be done during next week
> 
> Then let's wait for it. @shimmybalsam can you ping me here when that is done?

Hi @jirikraus you can find the content of updated version (waiting to be approved and added into wiki by manju) in this issue i just opened: https://github.com/openucx/ucc/issues/726
Let me know if you have any further questions or comments.

Sergei-Lebedev commented:
hi @jirikraus, both PRs for config file and log improvement were merged

jirikraus commented on docs/user_guide.md:
````diff
@@ -306,6 +308,36 @@ ucc_coll_score_map.c:185  UCC  INFO       RocmManaged: {0..4095}:TL_UCP:10 {4K..
 [...] snip
 ```
 
+The next release of UCC add the `UCC_COLL_TRACE` environment variable:
````
@Sergei-Lebedev should we be more specific here?

Sergei-Lebedev replied:
yes, UCC 1.2.0 to be precise

jirikraus replied:
Thanks @Sergei-Lebedev I have update to

UCC 1.2.0 or newer supports the `UCC_COLL_TRACE` environment variable:

resolving this.

jirikraus commented:
@manjugv can this be merged?

Sergei-Lebedev commented:
> @manjugv can this be merged? The two checks that fail timed out so I doubt there is anything I can do about this.

merged, thanks @jirikraus
